core/src/main/java/hudson/model/AbstractCIBase.java

```diff
@@ -197,15 +199,20 @@ public void run() {
                 for (Computer c : old) {
                     c.inflictMortalWound();
                 }
+
+                getQueue().scheduleMaintenance();
             }
-        });
-        for (Computer c : old) {
-            // when we get to here, the number of executors should be zero so this call should not need the Queue.lock
-            killComputer(c);
-        }
-        getQueue().scheduleMaintenance();
-        for (ComputerListener cl : ComputerListener.all())
-            cl.onConfigurationChange();
+        }).addListener(new Runnable() {
+            @Override
+            public void run() {
+            for (Computer c : old) {
+                // when we get to here, the number of executors should be zero so this call should not need the Queue.lock
+                killComputer(c);
+            }
+                for (ComputerListener cl : ComputerListener.all())
+                    cl.onConfigurationChange();
+            }
+        }, MoreExecutors.sameThreadExecutor());
```

Comment:
WAT! You have just moved this back inside Queue.withLock() because the `addListener(..., MoreExecutors.sameThreadExecutor())` will just be executing on Queue.mutator also and thus we have the very problem that you complained about and I introduced the `kill/inflictMortalWound` split to handle... 

This is wrong